### DISCUSSION TITLE
Inline::Javascript as a Raku grammar

### DISCUSSION THREAD

**Discussion by lizmat** [Raku/problem-solving#402] [Ideas]
https://github.com/Raku/problem-solving/discussions/395#discussioncomment-7557519 brought up the idea of having an Inline::Javascript **grammar** that would convert a Javascript program / library to a RakuAST tree, and thus be able to execute Javascript as native raku code.

It was suggested to give this its own discussion.  This is it.

**Comment by lizmat** [comment]
Looking at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types, it looks like a fair start would be to take the current Raku grammar / actions, strip it of unsupported syntax, and start adding/adapting Javascript specific tokens / rules / methods.

**Comment by niner** [comment]
I don't think this is really a Raku problem to solve. Sounds like a module. Especially given the stated goal of interoperability with existing Javascript libraries. Parsing is by far the easiest part. The hard part is the runtime semantics. None of the existing Raku building blocks will be re-usable for that. Neither the RakuAST nodes, nor anything at all from the setting.

**Comment by lizmat** [comment]
> I don't think this is really a Raku problem to solve. Sounds like a module

Very likely.  But is that a reason to *not* discuss this here?  If it is, where else?

> The hard part is the runtime semantics.

Yes, I know.  Still I think it would have better chance than a slang for Perl, as JS doesn't have XS, right?

> None of the existing Raku building blocks will be re-usable for that. Neither the RakuAST nodes, nor anything at all from the setting.

I got the impression that it actually *would* be possible to re-use a large part of it.  Would you care to elaborate on why it wouldn't be possible?

**Comment by niner** [comment]
> Very likely. But is that a reason to _not_ discuss this here? If it is, where else?

A lot of people are watching this repo because they want to participate in topics about Raku's language design or community issues but may not be interested in how to write module X.

> > The hard part is the runtime semantics.
> 
> Yes, I know. Still I think it would have better chance than a slang for Perl, as JS doesn't have XS, right?
> 
> I got the impression that it actually _would_ be possible to re-use a large part of it. Would you care to elaborate on why it wouldn't be possible?

Just about all semantics you could think of are different between Javascript and Raku. E.g. variables in Javascript are function scoped while in Raku we are as lexical as possible. Everything about OO is different including method resolution. You'd think there is a lot of common ground, e.g. most basic things like simple addition of numbers, right? [No, it isn't](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-ecmascript-language-types-number-type). Same goes for string handling. On a first glance it's just the same. I mean string concatenation is string concatenation, right? But then you get into Unicode details and you start reimplementing all our primitives with no help from the VM.

It's easy to get something that looks and feels some 80 % like Javascript. That's often good enough if you e.g. want to provide some scripting interface to your language, i.e. users writing simple scripts using your APIs. But for real interop you need to nail down 100 % and that means even supporting the differences in NaN handling between Javascript and IEEE 754-2019.

**Comment by lizmat** [comment]
Closing as Off-Topic here.
